features/support/utils_gen.rb

```diff
@@ -22,67 +20,80 @@ def set_current_user(user = nil)
 def embed_image(name, take_screenshot = true)
   screenshot_path = File.dirname(__FILE__) + '/media/' + name + '.png'
 
-  if take_screenshot
-    screenshot(screenshot_path)
-  end
+  screenshot(screenshot_path) if take_screenshot
+
   embed(screenshot_path, "image/png", "SCREENSHOT")
 end
 
-
 def embed_screenshot scenario
   name = scenario.name.gsub(" ", "_").downcase
   screenshot_path = File.dirname(__FILE__) + '/media/' + ENV['platform'] + '/'
   screenshot_embed(options = {:prefix => screenshot_path, :name => name})
 end
 
 def wait_until_loader_disappear
-  while query("* id:'wait_spinner'").size != 0
-    sleep 0.5
+  begin
+    Timeout::timeout(30) do
+      while query("* id:'wait_spinner'").size != 0
+        sleep 0.5
+      end
+    end
+  rescue Timeout::Error
+    raise "Loader doesn't disappear"
   end
 end
 
 def wait_until_text_disappear(text)
-  while check_if_text_exists(text)
-    sleep 0.5
+  Timeout::timeout(30) do
+    while check_if_text_exists(text)
+      sleep 0.5
+    end
   end
 end
 
 def wait_until_text_visible(text)
-  wait_until_element_ready "* {text CONTAINS[C] '#{text}'}"
+  wait_until_element_visible "* {text CONTAINS[C] '#{text}'}"
 end
 
-def wait_until_element_ready(qr)
-  10.times do
-    if query(qr).size > 0
-      break
-    else
-      sleep 1
-      puts "- Element is not ready yet ...."
-    end
+def wait_until_element_visible(qr)
+  begin     
+    Timeout::timeout(30) do
+      while !element_exists(qr)
+        sleep 1
+      end
+    end  
+  rescue Timeout::Error
+    raise 'Element is not visible.'
   end
 end
 
-def wait_until_one_element_ready(*args)
-  # this will wait until one of the element visible in the given query list
-  # wait_until_one_element_ready("* text:'Stores'", "* text:'Places'")
+def wait_until_one_element_visible(*args)
+  # this waits until one of the element visible in the given query list
+  # wait_until_one_element_visible("* text:'Stores'", "* text:'Places'")
 
   element = :not_exist
 
-  10.times do
+  30.times do
     args.each do |qr|
-      if query(qr).size > 0
+      if element_exists(qr)
+        # set the element exist and leave the each loop 
         element = :exist
         break
       end
     end
 
     if element == :exist
+      # leave the times loop if exist
       break
     else
       sleep 1
-      puts "- Element is not ready yet ...."
     end
   end
+
+  if element == :not_exist
```

Comment:
you don't need a clause around this raise if you will use `return` instead of `break` in the loop above. this way the function will just return when found element and in the bad case scenario raise if it didn't

Reply:
thank you clarification,  this is better: 

```
def wait_until_one_element_visible(*args)
  # this waits until one of the element visible in the given query list
  # wait_until_one_element_visible("* text:'Stores'", "* text:'Places'")

  30.times do
    args.each do |qr|
      return true if element_exists(qr)
    end

    sleep 1
  end

  raise '- None of the elements is visible'
end
```